keps/sig-node/3695-pod-resources-for-dra/README.md

```diff
@@ -0,0 +1,372 @@
+
+## Motivation
+
+The PodResources API is currently being used by CNI meta-plugins like [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [DANM](https://github.com/nokia/danm) to add resources allocated by device plugins as CNI arguments. One of the primary motivations for this KEP is to extend the PodResources API to allow these CNI plugins to reference resources allocated by DRA as well. Additional extensions to API will also make it easier for CNI meta-plugins to access resources allocated to a specific pod, rather than having to filter through resources for all pods on the node. Another use-case is to allow node monitoring agents to access information about allocated resources by DRA as well.
+
+### Goals
+
+Enabling the DRA feature to work with CNIs that require complex network devices such as RDMA. DRA resource drivers will allocate the resources, and the meta-plugin will read the allocated [CDI Device](https://github.com/container-orchestrated-devices/container-device-interface) using the Pod resource API. The meta-plugin will then inject the device-id as CNI arguments and invoke other CNIs (just as it does for devices allocated by the device plugin today).
+
+### Non-Goals
+
+To enhance the GetAllocatableResources() call in the PodResources API to account for resources managed by DRA. With DRA there is no standard way to get the capacity, for example.
+
+## Proposal
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks. The aspects that we consider below are as follows:
+- What are the risks associated with the API service itself?
+- What are the risks associated with the data itself?
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests risk impacting the kubelet performances | High          | Implement rate limiting and or passive caching, follow best practices for gRPC resource management. |
+| Improper access to the data | Low | Server is listening on a root owned unix socket. This can be limited with proper pod security policies. |
+
+## Design Details
+
+### Proposed API
+
+
+Our proposal is to extend the existing pod resources gRPC service of the Kubelet by adding the CDIDevices field to the ContainerResources message. This new field will contain information about the DRA driver name, the DRA resource class, and a list of CDI Devices allocated by that DRA driver. It's important to note that this proposal focuses on retrieving CDI devices allocated by DRA. Other mechanisms that may utilize CDI, such as device plugin, are not included in the scope of this proposal.
```

Comment:
Also, since this all seems closely tied to DRA (and not just CDI), I don't think the top-level field should be called `CDIDevices`, but rather something like `DynamicResources`  (with an embedded field for CDIDevices alongside the other fields mentioned here).

I'm open to an alternate name, but I don't think `CDIDevices` is appropriate here.

Reply:
> We should be clear about what we mean by `DriverName` here. There are two pieces to a driver -- the `DriverName` refers to the name of the controller which does allocations, and the `KubeletPluginName` refers to the kubelet plugin which does the actual allocation of CDI devices.
> 
> At present they are one and the same (i.e. each controller is tied in a 1:1 relationship with its kubelet plugin).
> 
> However, we are in the process of decoupling these so that a single controller can trigger resource allocations from multiple kubelet plugins (which may or may not have the same name as the `DriverName` of the controller).
> 
> So which bit is actually important for you here? The name of the controller that did the allocation, or the name of the kubelet plugin that generated the CDI device list?

After rethinking this, because the resource class is cluster scope  I don't need the DriverName. The name of the resource class is  unique in the cluster so we can map the network attachment definition to the resource class. I removed the DriverName from the proto.

Reply:
Beware that the content of the resource class and thus the driver name there can change over the lifetime of a resource claim. We cannot prevent delete+recreate.